File: android/src/main/java/com/trustwalletcorereactnative/wallet_core/coin/Binance.kt
package com.trustwalletcorereactnative.wallet_core.coin

import com.google.protobuf.ByteString
import com.trustwalletcorereactnative.wallet_core.error.CEError
import com.trustwalletcorereactnative.wallet_core.error.CError
import com.trustwalletcorereactnative.wallet_core.util.toHex
import com.trustwalletcorereactnative.wallet_core.util.toHexByteArray
import com.trustwalletcorereactnative.wallet_core.util.toLong
import wallet.core.java.AnySigner
import wallet.core.jni.AnyAddress
import wallet.core.jni.CoinType
import wallet.core.jni.PrivateKey
import wallet.core.jni.proto.Binance

class Binance : Coin(CoinType.BINANCE, "m/44'/714'/0'/0/0") {

    override fun signTransaction(tx: Map<String, Any>, privateKey: String): String {
        val transaction: Map<String,Any> = tx["transaction"] as? Map<String,Any> ?: throw CError(
            CEError.MissingArgumentError.message,
            CEError.MissingArgumentError.code,)

        val publicKey = PrivateKey(privateKey.toHexByteArray()).getPublicKeySecp256k1(true)
        val signingInput = Binance.SigningInput.newBuilder().apply {
            chainId = tx["chainId"] as String
            accountNumber = tx["accountNumber"]!!.toLong()
            sequence = tx["sequence"]!!.toLong()
            source = tx["source"]!!.toLong()
            memo = tx["memo"] as String
            this.privateKey = ByteString.copyFrom(privateKey.toHexByteArray())
        }


        val token = Binance.SendOrder.Token.newBuilder().apply {
            denom = transaction["denom"] as String
            amount = transaction["amount"]!!.toLong()
        }

        val input = Binance.SendOrder.Input.newBuilder()
        input.address = ByteString.copyFrom(AnyAddress(publicKey, coinType).data())
        input.addAllCoins(listOf(token.build()))

        val output = Binance.SendOrder.Output.newBuilder()
        output.address =
            ByteString.copyFrom(AnyAddress(transaction["to"] as String, coinType).data())
        output.addAllCoins(listOf(token.build()))

        val sendOrder = Binance.SendOrder.newBuilder()
        sendOrder.addAllInputs(listOf(input.build()))
        sendOrder.addAllOutputs(listOf(output.build()))

        signingInput.sendOrder = sendOrder.build()

        val signed: Binance.SigningOutput =
            AnySigner.sign(signingInput.build(), coinType, Binance.SigningOutput.parser())


        return signed.encoded.toByteArray().toHex()
    }

}

File: ios/WalletCore/Coin/Binance.swift
//
//  Binance.swift
//  trust-wallet-core-react-native
//
//  Created by Nguyen Van Toan on 21/5/24.
//

import Foundation
import WalletCore

class Binance : Coin{
    init() {
        super.init(coinType: CoinType.binance, derivationPath: "m/44'/714'/0'/0/0")
    }
    
    override func signTransaction(privateKey : String, tx: Dictionary<String, Any>) throws -> String {
        
        let transaction = tx["transaction"] as? Dictionary<String,Any>
        
        guard let transaction = transaction else {
            throw CError.MissingArgumentError
        }
        
        guard let pvData = Data(hexString: privateKey) else {
            throw CError.InvalidHexArgumentError
        }
        
        guard let pv = PrivateKey(data: pvData) else {
            throw CError.Unknown
        }
        
        let publicKey = pv.getPublicKeySecp256k1(compressed: true)
        
        let token = BinanceSendOrder.Token.with {
            $0.denom = transaction["denom"] as! String
            $0.amount = transaction["amount"] as! Int64
        }
        
        let orderInput = BinanceSendOrder.Input.with {
            $0.address = AnyAddress(publicKey: publicKey, coin: coinType).data
            $0.coins = [token]
        }
        
        let orderOutput = BinanceSendOrder.Output.with {
            $0.address = AnyAddress(string: transaction["to"] as! String, coin: coinType)!.data
            $0.coins = [token]
        }
        
        var input = BinanceSigningInput.with{
            $0.chainID = tx["chainId"] as! String
            $0.memo = tx["memo"] as! String
            $0.accountNumber = tx["accountNumber"] as! Int64
            $0.source = tx["source"] as! Int64
            $0.sequence = tx["sequence"] as! Int64
            $0.privateKey = pvData
            $0.sendOrder = BinanceSendOrder.with{
                $0.inputs = [orderInput]
                $0.outputs = [orderOutput]
            }
        }
        
        let signed : BinanceSigningOutput = AnySigner.sign(input: input, coin: coinType)
        
        return signed.encoded.hexString
    }
}
